extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/AbstractJpaOperations.java

```diff
@@ -0,0 +1,506 @@
+package io.quarkus.hibernate.reactive.panache.common.runtime;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import javax.enterprise.util.TypeLiteral;
+import javax.persistence.LockModeType;
+import javax.persistence.PersistenceException;
+import javax.transaction.SystemException;
+import javax.transaction.TransactionManager;
+
+import org.hibernate.internal.util.LockModeConverter;
+import org.hibernate.reactive.mutiny.Mutiny;
+import org.hibernate.reactive.mutiny.Mutiny.Session;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.panache.common.Parameters;
+import io.quarkus.panache.common.Sort;
+import io.quarkus.panache.common.exception.PanacheQueryException;
+import io.smallrye.mutiny.Multi;
+import io.smallrye.mutiny.Uni;
+
+public abstract class AbstractJpaOperations<PanacheQueryType> {
+
+    protected abstract PanacheQueryType createPanacheQuery(Uni<Mutiny.Session> em, String query, String orderBy,
+            Object paramsArrayOrMap);
+
+    protected abstract Uni<List<?>> list(PanacheQueryType query);
+
+    protected abstract Multi<?> stream(PanacheQueryType query);
+
+    //
+    // Instance methods
+
+    public Uni<Void> persist(Object entity) {
+        return getEntityManager().flatMap(em -> persist(em, entity));
+    }
+
+    public Uni<Void> persist(Mutiny.Session em, Object entity) {
+        if (!em.contains(entity)) {
```

Comment:
What is the purpose of this check? And why should it not be persisted otherwise?

Reply:
Also, why is this specific variation of the `persist` method having this check, but the other methods won't have the same semantics?

Reply:
The purpose is to not call `persist` if it's already persisted. The ORM version has the same check. What happens if I call `persist` on an already persisted entity, otherwise?

Reply:
The other variations of `persist` ultimately call this, so it's fine.

Reply:
>  What happens if I call persist on an already persisted entity, otherwise?

Nothing, but at least we have a chance to verify the transaction is alive, extend timeouts, and things like that.

Reply:
Well, the javadoc wasn't entirely clear about the semantics here.